src/internal/pfsdb/common.go

```diff
@@ -0,0 +1,184 @@
+	Field  T
+	Value  any
+	values []any
+	Op     FilterOperator
+}
+
+func (f *Filter[T]) String() (formatted string, err error) {
+	formatted = fmt.Sprintf(string(f.Op), f.Field)
+	if f.Op == ValueIn {
+		formatted, f.values, err = sqlx.In(formatted, f.Value) // Thank god for this!
+		if err != nil {
+			return "", errors.Wrapf(err, "expanding %s", formatted)
+		}
+	}
+	return
+}
+
+type AndFilters[T FieldType] []*Filter[T]
+
+func (f AndFilters[T]) String() (string, error) {
+	var parts []string
+	for _, filter := range f {
+		part, err := filter.String()
+		if err != nil {
+			return "", err
+		}
+		parts = append(parts, part)
+	}
+	return strings.Join(parts, " AND "), nil
+}
+
+type QueryBuilder[T FieldType] struct {
+	baseQuery   string
+	queryParams []any
+
+	AndFilters AndFilters[T]
+	GroupBy    string
+	OrderBy    string
+	// Limit, Offset uint64
+}
+
+func (c *QueryBuilder[T]) Build() (string, error) {
+	query := c.baseQuery
+	condition, err := c.AndFilters.String()
+	if err != nil {
+		return "", err
+	}
+	for _, filter := range c.AndFilters {
+		if filter.Op == ValueIn {
+			// filter.values is set by Filter.String()
+			c.queryParams = append(c.queryParams, filter.values...)
+		} else {
+			c.queryParams = append(c.queryParams, filter.Value)
+		}
+	}
+	if condition != "" {
+		query += "\nWHERE " + condition
+	}
+	// TODO implement GROUP BY and ORDER BY in a type safe way
+	if c.GroupBy != "" {
+		query += "\nGROUP BY " + c.GroupBy
+	}
+	if c.OrderBy != "" {
+		query += "\nORDER BY " + c.OrderBy
+	}
+	query = sqlx.Rebind(sqlx.DOLLAR, query)
+	return query, nil
+}
+
+type transformFn[T ModelType, U PairType] func(context.Context, *pachsql.Tx, *T) (*U, error)
+
+// T is the type used to deserialize rows from the database
+// U is the type the type returned by the iterator
+// Typically the user supplies the transform function, which converts T to U
+type pageIterator[T ModelType, U PairType] struct {
+	db            *pachsql.DB
+	baseQuery     string
+	queryParams   []any
+	limit, offset uint64
+	page          []*U
+	pageIndex     int
+	transform     transformFn[T, U] // converts T to U, and is called for each item in the page
+}
+
+func newPageIterator[T ModelType, U PairType, S FieldType](ctx context.Context, db *pachsql.DB, qb QueryBuilder[S], transform transformFn[T, U], limit, offset uint64) (pageIterator[T, U], error) {
+	baseQuery, err := qb.Build()
+	if err != nil {
+		return pageIterator[T, U]{}, err
+	}
+	iter := pageIterator[T, U]{
+		db:          db,
+		baseQuery:   baseQuery,
+		queryParams: qb.queryParams,
+		limit:       limit,
+		offset:      offset,
+		transform:   transform,
+	}
+	return iter, nil
+}
+
+func (i *pageIterator[T, U]) nextPage(ctx context.Context) error {
+	query := i.baseQuery + fmt.Sprintf("\nLIMIT %d OFFSET %d", i.limit, i.offset)
+	var page []*U
+	// open a transaction and get the next page of results
+	if err := dbutil.WithTx(ctx, i.db, func(ctx context.Context, tx *pachsql.Tx) error {
```

Comment:
Do you want to open a new transaction for the next page of results? I think for most resources you can probably reuse the transaction.